src/lib.rs

```diff
@@ -146,6 +153,17 @@ fn blake3(_: Python, m: &PyModule) -> PyResult<()> {
             )
         }
 
+        /// Return a copy of a Blake3Hasher object.
+        /// The usual caveats of multithreading as mentioned in `./README.md` apply here.
+        /// Calling `copy` in a multi-threaded situation without a lock on the copied object
+        /// will likely result in non-deterministic behavior.
```

Comment:
> Calling `copy` in a multi-threaded situation without a lock on the copied object will likely result in non-deterministic behavior.

I think you could say something similar about almost any object in Python. In fact, I think you could say something stronger than that. If you're calling methods on a single object from multiple threads without locking, you're very likely to get _nonsense_ results. Not just nondeterministic results, like two things happening in an unknown order, but truly nonsensical results, like basic invariants being violated. For example (forgive me if this is already clear to you), this Python function is incorrect under multithreading:

```python
def increment_first(int_list):
    int_list[0] += 1
```

If two different threads call that method on the same list, one of the increments might get dropped, because `+=` is not atomic. This is a nonsense result, which a single threaded program could never produce. Though to be fair, it's not "undefined behavior" in the C/Rust sense. The interpreter isn't allowed to produce a _random_ result, or to corrupt some arbitrary piece of memory. The GIL is how the Python interpreter makes sure it doesn't do those things, even when it's producing nonsense.

This difference is what the readme is getting at when it talks about thread safety and undefined behavior. Because the `update` method explicitly releases the GIL, it's vulnerable to the C/Rust style of undefined behavior, in a way that most Python code is not. Technically this is an important difference. We might actually corrupt arbitrary memory here. (That's not likely in this case, but our languages are no longer giving us any guarantees, and we're forced to reason about things like typical compiler optimizations and processor-specific memory behavior. That's never where we want to be.) However, in practice, code that triggers this was probably in a bad place to begin with. We're talking about the difference between "nonsense" and "worse than nonsense". Like, you already have a terrible correctness issue, and now we're telling you that it might be a crasher or a security issue too. That's interesting, but either way the end result is that you need to use a mutex.

So anyway, does this function merit a special warning about concurrency? I'm not sure. My leaning is that it's kind of par for the course in Python. The GIL-related differences are interesting to experts, but not relevant for folks writing non-pathological application code. And notably, the design decisions we're making here are the same as what `hashlib` does in the Python standard library.

Reply:
> We're talking about the difference between "nonsense" and "worse than nonsense".

Ahh I see what you are getting at now. Yeah I'll update the comment since this seems to be a different issue. 

TBH, and I think you're aware of this, the thread safety thing should probably just be addressed at the wrapper level. I was able to get some weird effects even with a threading.Lock.

Reply:
What do you mean by the wrapper level? Do you mean like the PyO3 level? They've done quite a lot to guarantee safety, and the issue is more that we're explicitly using unsafe code and pointer casting to disable the safeguards they put in place. Releasing the GIL, while reading a memory buffer that other Python threads could have a referece to, is just fundamentally unsound. No safe abstraction could allow it. But that's what the Python standard library does with hashing, and that's what we need to do too 😬 

At a high level, Rust is able to achieve a safer story here because it enforces the uniqueness of `&mut` references at every single point where they're used. But if mutable aliasing is allowed in even a single place -- and most languages allow it everywhere -- it's hard to make any guarantees. Python code is in no position to say "no one is capable of writing to this value." All it can say is "you have the GIL." When we don't want to hold the GIL, we're opting out of all the safety Python can provide.

Reply:
Yeah, nevermind. I was seeing weird nonreproducible behavior when I was profiling and thought it might be related to this issue, but after a bit of digging it seems these all have to do with readinto(). There's definitely some nuance to using it vs the conventional read.